File: iOS/Crisis Connect/Features/Messaging/Signal/SignalPreKeyManager.swift
//
//  SignalPreKeyManager.swift
//  Crisis Connect
//
//  FS-6: generates, locally persists, and publishes the device's Signal-protocol prekeys — the
//  iOS port of Android's `SignalPreKeyManager`, speaking to the same `publishSignalPreKeys` /
//  `checkSignalPreKeys` callables. Every key handed to the backend is stored in the protocol store
//  FIRST so a peer's session opened against it can always be decrypted (the backend copy is
//  public-only).
//

import Foundation
import LibSignalClient

final class SignalPreKeyManager {
    private let store: SignalProtocolFileStore
    private let client: InternetMessagingClient
    private static let defaults = UserDefaults.standard

    private static let keyPreKeyId = "cc.signal.next_prekey_id"
    private static let keySignedId = "cc.signal.next_signed_prekey_id"
    private static let keyKyberId = "cc.signal.next_kyber_prekey_id"
    private static let keyLastResortKyberId = "cc.signal.last_resort_kyber_id"
    private static let maxKeyId: Int = 0xFFFFFF
    private static let targetPool = 100
    private static let lowWatermark = 20
    private static let idLock = NSLock()

    init(store: SignalProtocolFileStore, client: InternetMessagingClient = InternetMessagingClient()) {
        self.store = store
        self.client = client
    }

    /// The published last-resort Kyber prekey id — the one `markKyberPreKeyUsed` must never delete.
    static func lastResortKyberId() -> UInt32 {
        UInt32(defaults.integer(forKey: keyLastResortKyberId))
    }

    /// Publish/rotate meta (identity, signed prekey, last-resort Kyber) and top up the one-time
    /// pools when the server reports them low. Mirrors Android's `ensurePublished`; call from the
    /// messaging bootstrap, off the main thread, best-effort.
    func ensurePublished() async throws {
        // Reinstall-reset recovery: iOS keeps the Signal identity in the keychain (survives app
        // deletion) but this prekey store lives in Application Support (wiped on deletion). A device
        // that lost its store but kept its identity keeps serving peers stale, undecryptable prekeys
        // — every inbound v3 message fails with missingRecord(<kyber id>) forever, because the
        // backend only clears its pool on an identity change. Detect it (local counters are fresh yet
        // the server still reports our prekeys published) and rotate the identity so the publish below
        // wipes the stale server pool. `keySignedId == 0` is only true on a genuinely fresh store —
        // an existing consistent install always has a non-zero counter, so normal users never rotate.
        let localFresh = Self.defaults.integer(forKey: Self.keySignedId) == 0
        if localFresh, let inv = try? await client.checkSignalPreKeys(), inv.published == true {
            NSLog("SignalPreKeyManager: fresh store + server has stale prekeys -> rotating identity")
            SignalIdentity.shared.rotate()
        }

        let identityPair = try SignalIdentity.shared.identityKeyPair()
        let context = NullContext()

        let signed = try generateSignedPreKey(identityPair)
        try store.storeSignedPreKey(signed, id: signed.id, context: context)
        let lastResort = try generateKyberPreKey(identityPair)
        try store.storeKyberPreKey(lastResort, id: lastResort.id, context: context)
        Self.defaults.set(Int(lastResort.id), forKey: Self.keyLastResortKyberId)

        let inventory = try? await client.checkSignalPreKeys()
        let needEc = Self.topUpCount(inventory?.ecCount, published: inventory?.published)
        let needKyber = Self.topUpCount(inventory?.kyberCount, published: inventory?.published)

        var oneTimeEc: [PreKeyRecord] = []
        for _ in 0..<needEc {
            let record = try generateOneTimePreKey()
            try store.storePreKey(record, id: record.id, context: context)
            oneTimeEc.append(record)
        }
        var oneTimeKyber: [KyberPreKeyRecord] = []
        for _ in 0..<needKyber {
            let record = try generateKyberPreKey(identityPair)
            try store.storeKyberPreKey(record, id: record.id, context: context)
            oneTimeKyber.append(record)
        }

        try await client.publishSignalPreKeys(
            registrationId: Int(SignalIdentity.shared.registrationId()),
            identityKeyBase64: identityPair.publicKey.serialize().base64EncodedString(),
            signedPreKey: PreKeyUpload(
                keyId: Int(signed.id),
                publicKeyBase64: try signed.publicKey().serialize().base64EncodedString(),
                signatureBase64: signed.signature.base64EncodedString()
            ),
            lastResortKyberPreKey: try Self.kyberUpload(lastResort),
            preKeys: try oneTimeEc.map { record in
                PreKeyUpload(
                    keyId: Int(record.id),
                    publicKeyBase64: try record.publicKey().serialize().base64EncodedString(),
                    signatureBase64: nil
                )
            },
            kyberPreKeys: try oneTimeKyber.map { try Self.kyberUpload($0) }
        )
    }

    private static func kyberUpload(_ record: KyberPreKeyRecord) throws -> PreKeyUpload {
        PreKeyUpload(
            keyId: Int(record.id),
            publicKeyBase64: try record.keyPair().publicKey.serialize().base64EncodedString(),
            signatureBase64: record.signature.base64EncodedString()
        )
    }

    private static func topUpCount(_ serverCount: Int?, published: Bool?) -> Int {
        let current = (published == true) ? (serverCount ?? 0) : 0
        return current >= lowWatermark ? 0 : max(targetPool - current, 0)
    }

    private func generateSignedPreKey(_ identity: IdentityKeyPair) throws -> SignedPreKeyRecord {
        let keyPair = PrivateKey.generate()
        let signature = identity.privateKey.generateSignature(message: keyPair.publicKey.serialize())
        return try SignedPreKeyRecord(
            id: Self.nextId(Self.keySignedId),
            timestamp: UInt64(Date().timeIntervalSince1970 * 1000),
            privateKey: keyPair,
            signature: signature
        )
    }

    private func generateKyberPreKey(_ identity: IdentityKeyPair) throws -> KyberPreKeyRecord {
        let keyPair = KEMKeyPair.generate()
        let signature = identity.privateKey.generateSignature(message: keyPair.publicKey.serialize())
        return try KyberPreKeyRecord(
            id: Self.nextId(Self.keyKyberId),
            timestamp: UInt64(Date().timeIntervalSince1970 * 1000),
            keyPair: keyPair,
            signature: signature
        )
    }

    private func generateOneTimePreKey() throws -> PreKeyRecord {
        try PreKeyRecord(id: Self.nextId(Self.keyPreKeyId), privateKey: PrivateKey.generate())
    }

    /// Monotonic id in [1, 0xFFFFFF] (libsignal's Medium range), wrapping to 1 — same as Android.
    private static func nextId(_ key: String) -> UInt32 {
        idLock.lock(); defer { idLock.unlock() }
        let next = defaults.integer(forKey: key) + 1
        let wrapped = (1...maxKeyId).contains(next) ? next : 1
        defaults.set(wrapped, forKey: key)
        return UInt32(wrapped)
    }
}

/// Wire shape of one uploaded prekey — matches Android `PreKeyUpload.toMap()`.
struct PreKeyUpload {
    let keyId: Int
    let publicKeyBase64: String
    let signatureBase64: String?

    func toMap() -> [String: Any] {
        var m: [String: Any] = ["keyId": keyId, "publicKey": publicKeyBase64]
        if let signatureBase64 { m["signature"] = signatureBase64 }
        return m
    }
}

File: Android/app/src/main/java/com/auralis/crisisconnect/messaging/signal/SignalPreKeyManager.kt
package com.auralis.crisisconnect.messaging.signal

import android.content.Context
import android.util.Base64
import com.auralis.crisisconnect.data.AppDatabase
import com.auralis.crisisconnect.messaging.InternetMessagingClient
import com.auralis.crisisconnect.security.KeystoreBackedPreferences
import org.signal.libsignal.protocol.ecc.ECKeyPair
import org.signal.libsignal.protocol.kem.KEMKeyPair
import org.signal.libsignal.protocol.kem.KEMKeyType
import org.signal.libsignal.protocol.state.KyberPreKeyRecord
import org.signal.libsignal.protocol.state.PreKeyRecord
import org.signal.libsignal.protocol.state.SignedPreKeyRecord

/**
 * Generates, locally persists, and publishes the device's Signal-protocol prekeys (FS-2). Every key
 * it hands to the backend is first stored in the [AndroidSignalProtocolStore] so that when a peer
 * consumes it and opens a session against us, we can decrypt (the backend copy is public-only).
 *
 * Ownership of ids: monotonically increasing counters kept in a Keystore-encrypted pref so ids never
 * repeat across top-ups. The signed prekey and last-resort Kyber prekey are rotated in full each
 * time we publish meta (≈ on every bootstrap / rotation); one-time pools are topped up when the
 * server reports them low.
 */
class SignalPreKeyManager(
    context: Context,
    private val store: AndroidSignalProtocolStore,
    private val identity: SignalIdentity,
    private val client: InternetMessagingClient = InternetMessagingClient(),
) {
    private val appContext = context.applicationContext
    private val counters = KeystoreBackedPreferences(appContext, PREFS_NAME, STORE_KEY_ALIAS)

    /**
     * Ensure the device has published Signal prekeys and keeps its one-time pools stocked. Called
     * from bootstrap. Republishes meta (identity/signed/last-resort Kyber) every time — cheap and
     * doubles as signed-prekey rotation — then tops up the one-time pools if the server says they're
     * below [LOW_WATERMARK].
     */
    suspend fun ensurePublished() {
        val identityKeyPair = identity.identityKeyPair()

        // Rotate the signed + last-resort Kyber prekeys and store them locally.
        val signedPreKey = generateSignedPreKey()
        store.storeSignedPreKey(signedPreKey.id, signedPreKey)
        val lastResortKyber = generateKyberPreKey()
        store.storeKyberPreKey(lastResortKyber.id, lastResortKyber)

        // How many one-time keys to publish now: fill to TARGET_POOL if the server is low (or empty).
        val inventory = runCatching { client.checkSignalPreKeys() }.getOrNull()
        val needEc = topUpCount(inventory?.ecCount, inventory?.published)
        val needKyber = topUpCount(inventory?.kyberCount, inventory?.published)

        val oneTimeEc = List(needEc) { generateOneTimePreKey() }
        oneTimeEc.forEach { store.storePreKey(it.id, it) }
        val oneTimeKyber = List(needKyber) { generateKyberPreKey() }
        oneTimeKyber.forEach { store.storeKyberPreKey(it.id, it) }

        client.publishSignalPreKeys(
            registrationId = identity.registrationId(),
            identityKeyBase64 = b64(identityKeyPair.publicKey.serialize()),
            signedPreKey = signedPreKey.toWire(),
            lastResortKyberPreKey = lastResortKyber.toWire(),
            preKeys = oneTimeEc.map { it.toWire() },
            kyberPreKeys = oneTimeKyber.map { it.toWire() },
        )
    }

    private fun topUpCount(serverCount: Int?, published: Boolean?): Int {
        // Server never published (published==false / null) → fill from scratch. Otherwise top up only
        // when below the low watermark, back to the target.
        val current = if (published == true) (serverCount ?: 0) else 0
        return if (current >= LOW_WATERMARK) 0 else (TARGET_POOL - current).coerceAtLeast(0)
    }

    private fun generateSignedPreKey(): SignedPreKeyRecord {
        val keyPair = ECKeyPair.generate()
        val signature = identity.identityKeyPair().privateKey.calculateSignature(keyPair.publicKey.serialize())
        return SignedPreKeyRecord(nextId(KEY_SIGNED_ID), System.currentTimeMillis(), keyPair, signature)
    }

    private fun generateKyberPreKey(): KyberPreKeyRecord {
        val keyPair = KEMKeyPair.generate(KEMKeyType.KYBER_1024)
        val signature = identity.identityKeyPair().privateKey.calculateSignature(keyPair.publicKey.serialize())
        return KyberPreKeyRecord(nextId(KEY_KYBER_ID), System.currentTimeMillis(), keyPair, signature)
    }

    private fun generateOneTimePreKey(): PreKeyRecord =
        PreKeyRecord(nextId(KEY_PREKEY_ID), ECKeyPair.generate())

    /** Monotonic id in [1, 0xFFFFFF]; wraps to 1 to stay inside libsignal's Medium range. */
    @Synchronized
    private fun nextId(counterKey: String): Int {
        val next = counters.getInt(counterKey, 0) + 1
        val wrapped = if (next in 1..MAX_KEY_ID) next else 1
        counters.putInt(counterKey, wrapped)
        return wrapped
    }

    private fun b64(bytes: ByteArray): String = Base64.encodeToString(bytes, Base64.NO_WRAP)

    private fun PreKeyRecord.toWire(): InternetMessagingClient.PreKeyUpload =
        InternetMessagingClient.PreKeyUpload(id, b64(keyPair.publicKey.serialize()), null)

    private fun SignedPreKeyRecord.toWire(): InternetMessagingClient.PreKeyUpload =
        InternetMessagingClient.PreKeyUpload(id, b64(keyPair.publicKey.serialize()), b64(signature))

    private fun KyberPreKeyRecord.toWire(): InternetMessagingClient.PreKeyUpload =
        InternetMessagingClient.PreKeyUpload(id, b64(keyPair.publicKey.serialize()), b64(signature))

    companion object {
        private const val PREFS_NAME = "crisisconnect_signal_prekey_counters"
        private const val STORE_KEY_ALIAS = "cc_signal_prekey_counters_v1"
        private const val KEY_PREKEY_ID = "next_prekey_id"
        private const val KEY_SIGNED_ID = "next_signed_prekey_id"
        private const val KEY_KYBER_ID = "next_kyber_prekey_id"
        private const val MAX_KEY_ID = 0xFFFFFF
        private const val TARGET_POOL = 100
        private const val LOW_WATERMARK = 20

        fun create(context: Context): SignalPreKeyManager {
            val identity = SignalIdentity(context)
            val store = AndroidSignalProtocolStore(
                dao = AppDatabase.getInstance(context).signalStoreDao(),
                identity = identity,
            )
            return SignalPreKeyManager(context, store, identity)
        }
    }
}
